### PR TITLE
Ignore expired silences OnGossip

### PR DESCRIPTION
**Pull request by josedonizetti**
This will fix the bug of resync deleted silences due to the state of other peers. More context on https://github.com/prometheus/alertmanager/issues/996

### CODE REVIEW

**Comment by stuartnelson3** [comment]
Thanks for the PR! I will have to think about this, as it opens up new issues: some people like to configure their data retention time high so that they can view and recreate silences from expired ones. By not gossiping expired silences, there will be inconsistency between the alertmanagers in a mesh.

**Comment by josedonizetti** [comment]
@stuartnelson3 

This isn't about ignoring gossiped expired silences, those will still exist in all HA, no inconsistency on this part. The PR ignores only GCed expired silences. I've started to test this after seen this issue https://github.com/prometheus/alertmanager/issues/996.

The test was done on current master. 

What I did:
1) decrease the data retention to a smaller window (5 min)
2) decrease the GC ticker to a smaller window (3 min)
3) start 3 instances of HA
4) send a few alerts
5) silence those alerts
6) expire those silences
7) wait to see if the GC would clean them

What I saw:

The GC would run on a specific instance 'A' and would clean the expired silences. But just a few seconds back the expired alerts would reappear in 'A' due to a gossiped of another instance 'B' or 'C', where the GC hadn't run.

Let me know if there more information I can gather to help out on this issue.

**Comment by beorn7** [comment]
@stuartnelson3 are you still up to reviewing this?

The not-happening GC is hurting us quite a lot by now, so I'm tempted to work on this myself, but I don't want to re-do work already done or work redundantly on something that is already being worked on behind the scenes.

**Comment by stuartnelson3** [comment]
Feel free to grab it, I was forced to have a weekend to myself and haven't touched it :)

File: silence/silence.go [hunk 1]
```diff
@@ -831,8 +832,8 @@ func (gd *gossipData) Merge(other mesh.GossipData) mesh.GossipData {
 	return gd
 }
 
-// mergeDelta behaves like Merge but returns a gossipData only
-// containing things that have changed.
+// mergeDelta behaves like Merge but ignore expired silences, and
```
Comment (stuartnelson3) [inline silence/silence.go]: `/s/ignore/ignores/`

File: silence/silence.go [hunk 2]
```diff
@@ -843,6 +844,11 @@ func (gd *gossipData) mergeDelta(od *gossipData) *gossipData {
 	defer gd.mtx.Unlock()
 
 	for id, s := range od.data {
+		// Only merge silence if not expired
```
Comment (stuartnelson3) [inline silence/silence.go]: Could you expand on this comment a bit? I know I'm going to come back to it in the future and need to refresh myself. Something along the lines of:

>If a gossiped silence is expired, skip it. For a given silence duration exceeding a few minutes, active silences will have already been gossiped. Once the active silence is gossiped, its expiration should happen more or less simultaneously on the different alertmanager nodes. Preventing the gossiping of expired silences allows them to be GC'd, and doesn't affect consistency across the mesh.

**Pull request merged**